packages/provider-utils/src/types/execute-tool.ts

```diff
@@ -0,0 +1,27 @@
+import { Tool, ToolCallOptions, ToolExecuteFunction } from './tool';
+import { isAsyncIterable } from '../is-async-iterable';
+
+export async function* executeTool<INPUT, OUTPUT>({
+  execute,
+  input,
+  options,
+}: {
+  execute: ToolExecuteFunction<INPUT, OUTPUT>;
+  input: INPUT;
+  options: ToolCallOptions;
+}): AsyncGenerator<
+  { type: 'preliminary'; output: OUTPUT } | { type: 'final'; output: OUTPUT }
+> {
+  const result = execute(input, options);
+
+  if (isAsyncIterable(result)) {
+    let lastOutput: OUTPUT | undefined;
+    for await (const output of result) {
+      lastOutput = output;
+      yield { type: 'preliminary', output };
+    }
+    yield { type: 'final', output: lastOutput! };
```

Comment:
from call with Lars: last output is doubled so that output shows immediately in the UI